### PR TITLE
[REF] web: load fontawesome through cdn

### PR DESCRIPTION
**Pull request by stefanorigano**
This commit reviews how fontAwesome is loaded by the system adding a
priority CDN option and using font-files in '/lib' as a fallback only.

It will also convert the original CSS file to SCSS, allowing to use a
SCSS function to output the necessary quite complicated src string.

FontAweome was made in 2017 and the original file included several
deprecated selector and vendor-prefix rules. These lines were no longer
necessary and have been removed.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by stefanorigano** [comment]
Hello ,

AL proposes to load fontawesome through a CDN, here a draft PR.
@JKE-be , I think it would be better to use our CDN ([eg](https://github.com/odoo-dev/odoo/blob/master-fontawesome-cdn-sri/addons/web/static/fonts/fonts.scss#L10)) rather than cloudflare, do you know how can I do that? 

cC. @ged-odoo , @odoo/rd-security

**Comment by mart-e** [comment]
What is the advantage of using a CDN instead of serving the CSS ourself? To save bandwidth? It's adding a new point of failure.

**Comment by stefanorigano** [comment]
> What is the advantage of using a CDN instead of serving the CSS ourself? To save bandwidth?

Hello @mart-e 
yes, save bandwidth, parallel connections...  I guess, it's definitively not my field ^^

>  It's adding a new point of failure.

We already do that [for other fonts](https://github.com/odoo-dev/odoo/blob/master-fontawesome-cdn-sri/addons/web/static/fonts/fonts.scss#L10)... do you think that could be a problem?

**Comment by JKE-be** [comment]
Hello

thanks for PR...

Strange, last month, @antonylesuisse ask to remove google font by default because you need to be able to launch Odoo in your cave without access to internet...
So if we use the cdn, we are going to make the opposite.

**Comment by mart-e** [comment]
> We already do that [for other fonts](https://github.com/odoo-dev/odoo/blob/master-fontawesome-cdn-sri/addons/web/static/fonts/fonts.scss#L10)... do you think that could be a problem? 

These were made for international fonts.  The files can take several MB and be needed only by few people (you rarely need to have both arabic and japanese fonts). Hosting them on our CDN was convenient and we did not need to put them in our git for people that may never need them.

One frequent argument of CDN for fonts/js is that it will be faster but that's not necesarily true with modern browsers bundling requests and so on ([example](https://wicki.io/posts/2020-11-goodbye-google-fonts/)).
Not that CDN are useless but not necesarily solving performance issues either.

**Comment by odony** [comment]
@antonylesuisse you'll have to make up your mind ;-)

FWIW @stefanorigano's proposal lists first the remote CDN urls and then the embedded ones, so it does fallback if the CDN isn't available. That takes care of the cave case, but doesn't help for the speed problem. For example there are places where the CDN URL will be slower than the local embedded one, or even blackholed due to network filters (e.g. the great firewall of China).

---

Quick comparison Runbot vs CDN: reloaded /web with no cache. The 77kB woff2 is downloaded at about the same time (around 1000ms in) and takes about 20-60ms in either case, either the CDN one or the embedded one.

**CDNJS**
Screenshot from 2021-12-22 23-38-20


**Embedded on Runbot**
Screenshot from 2021-12-22 23-30-08

However, when the CDN is blocked, the browser tries *3 different formats* before falling back to the embedded one, and *that alone* takes 100ms, with an instant block (using uBlock). I imagine this can cause multi-second delays on a slow/blocked network path:
Screenshot from 2021-12-22 23-32-16

In summary, it's not clear that this change can bring a measurable boost for anyone even in ideal situations, but there is clearly potential for a significant degradation of performance.

@stefanorigano we could change the order of the font-face src URLs and list alternatives in order of priorities, rather than having all the CDN URLs first. e.g. all woff2 URLs before the others, no?

PS: using CDNJS or our own fonts.odoocdn.com is an implementation choice that won't make any speed difference, in either case it will be served by CloudFlare using a "local" point of presence.

**Comment by stefanorigano** [comment]
@mart-e, @JKE-be, @odony , thanks for you feedbacks.

> However, when the CDN is blocked, the browser tries 3 different formats [...] we could change the order of the font-face src URLs and list alternatives in order of priorities

Indeed, fixed, thanks.
Also I changed the formats order,  we now prioritise the right ones.

| Order  | Format | Origin |
| ------------- | ------------- |  ------------- |
| 1  | woof2  | CDN  |
| 2  | woof2  |  `/lib` |
| 3  | woof | CDN  |
| 4  | woof  |  `/lib` |
| 5  | ttf  | CDN  |
| 6  | ttf |  `/lib` |
| 7  | eot  | CDN  |
| 8  | eot  |  `/lib` |
| 9  | svg  | CDN  |
| 10  | svg |  `/lib` |

> In summary, it's not clear that this change can bring a measurable boost for anyone even in ideal situations, but there is clearly potential for a significant degradation of performance.

Even if the previous "order fix" may solve the main performance issue, it's my understanding that CDN advantages are still debatable. I'll let you and @antonylesuisse decide ;)  

Btw there is still room for improvements, if not on the performance side at least for the footprint one.
Beside the CDN stuff I propose to:
1. Remove unnecessary fontAwesome css (-30 lines, already done in this PR)
2. Remove unnecessary fonts (~ 466 KB), keeping: 
   -`woff2` -> https://caniuse.com/?search=woff2
   -`woff` -> https://caniuse.com/?search=woff
   -`svg` (old iOs Devices) -> https://caniuse.com/?search=svg%20font
 
 Thanks again

**Comment by adr-odoo** [comment]
> @mart-e, @JKE-be, @odony , thanks for you feedbacks.
> 
> > However, when the CDN is blocked, the browser tries 3 different formats [...] we could change the order of the font-face src URLs and list alternatives in order of priorities
> 
> Indeed, fixed, thanks. Also I changed the formats order, we now prioritise the right ones.
> 
> Order	Format	Origin
> 1	woof2	CDN
> 2	woof2	`/lib`
> 3	woof	CDN
> 4	woof	`/lib`
> 5	ttf	CDN
> 6	ttf	`/lib`
> 7	eot	CDN
> 8	eot	`/lib`
> 9	svg	CDN
> 10	svg	`/lib`
> > In summary, it's not clear that this change can bring a measurable boost for anyone even in ideal situations, but there is clearly potential for a significant degradation of performance.
> 
> Even if the previous "order fix" may solve the main performance issue, it's my understanding that CDN advantages are still debatable. I'll let you and @antonylesuisse decide ;)
> 
> Btw there is still room for improvements, if not on the performance side at least for the footprint one. Beside the CDN stuff I propose to:
> 
> 1. Remove unnecessary fontAwesome css (-30 lines, already done in this PR)
> 2. Remove unnecessary fonts (~ 466 KB), keeping:
>    -`woff2` -> https://caniuse.com/?search=woff2
>    -`woff` -> https://caniuse.com/?search=woff
>    -`svg` (old iOs Devices) -> https://caniuse.com/?search=svg%20font
> 
> Thanks again

Feel free to remove svg if you want. We only support APple devices from iOS 12.5.5.

**Comment by stefanorigano** [comment]
Thanks so much to for your feedbacks.
This PR has been closed in favour of https://github.com/odoo/odoo/pull/83501,  I optimised fontAwesome bundle there 👍
